src/libcore/ffi.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Implementation of a `va_list`.
+
+#[cfg(any(all(not(target_arch = "aarch64"), not(target_arch = "powerpc"),
+              not(target_arch = "x86_64")),
+          windows))]
+#[cfg(not(stage0))]
+/// Basic byte pointer variant of a `va_list`.
+type VaListImpl = *const i8;
+
+#[cfg(all(target_arch = "aarch64", not(windows)))]
+#[cfg(not(stage0))]
+#[repr(C)]
+#[derive(Debug)]
+/// AArch64 ABI implementation of a `va_list`. See the
+/// [Aarch64 Procedure Call Standard] for more details.
+///
+/// [AArch64 Procedure Call Standard]:
+/// http://infocenter.arm.com/help/topic/com.arm.doc.ihi0055b/IHI0055B_aapcs64.pdf
+struct VaListImpl {
+    stack: *mut (),
+    gr_top: *mut (),
+    vr_top: *mut (),
+    gr_offs: i32,
+    vr_offs: i32,
+}
+
+#[cfg(all(target_arch = "powerpc", not(windows)))]
+#[cfg(not(stage0))]
+#[repr(C)]
+#[derive(Debug)]
+/// PowerPC ABI implementation of a `va_list`.
+struct VaListImpl {
+    gpr: u8,
+    fpr: u8,
+    reserved: u16,
+    overflow_arg_area: *mut (),
+    reg_save_area: *mut (),
+}
+
+#[cfg(all(target_arch = "x86_64", not(windows)))]
+#[cfg(not(stage0))]
+#[repr(C)]
+#[derive(Debug)]
+/// x86_64 ABI implementation of a `va_list`.
+struct VaListImpl {
+    gp_offset: i32,
+    fp_offset: i32,
+    overflow_arg_area: *mut (),
+    reg_save_area: *mut (),
+}
+
+/// A wrapper for a `va_list`
+#[cfg(not(stage0))]
+#[lang = "va_list"]
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct VaList<'a>(&'a mut VaListImpl);
+
+/// Trait wich limits the allowed types to be used with [VaList::arg]
+///
+/// [VaList::va_arg]: struct.VaList.html#method.arg
+pub trait VaArgSafe {}
```

Comment:
Created the `VaArgSafe` trait to whitelist primitive types.

Reply:
Can you make the trait private for now? I think that's what @rkruppe meant by "sealed trait".

Reply:
Updated

Reply:
The trait still appears to be public.

Reply:
I keep getting private trait in public interface errors when I do that. Is there another example of a "sealed trait" I could look at to get a hint about what I'm doing wrong here?

Reply:
Is that still an error? Or a warning, that the "deny warnings" we have set turns into an error? cc @petrochenkov

Reply:
Migration from errors to lints from https://github.com/rust-lang/rust/issues/48054 is not implemented yet :(
"Sealed trait" here would mean
```
mod my_private_module {
    pub trait VaArgSafe {} // Not reexported anywhere, available only inside the parent of my_private_module
}
```

Reply:
Yeah, that did the trick. Updated to use the suggested implementation of `VaArgSafe` and added a comment.